File: src/ServerFarming.Core/Services/Implement/PlantService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using FarmingDatabase.Model;
using ServerFarming.Core.Command;
using ServerFarming.Core.Model;
using ServerFarming.Core.Repositories;
using ServerFarming.Core.Exceptions;

namespace ServerFarming.Core.Services.Implement
{
    /// <summary>
    /// PlantService used for handing data related to a Plant before saving into Database
    /// </summary>
    public class PlantService : IPlantService
    {
        private readonly IPlantRepository plantRepository;
        private readonly IAuthenticationService authenticationService;
        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="plantRepository">used by DI</param>
        /// <param name="authenticationService">used by DI</param>
        public PlantService(IPlantRepository plantRepository,
            IAuthenticationService authenticationService)
        {
            this.plantRepository = plantRepository;
            this.authenticationService = authenticationService;
        }

        /// <summary>
        /// Add a Plant in a Farm Component
        /// </summary>
        /// <param name="farmingComponentDTO"></param>
        /// <param name="farmComponentId"></param>
        /// <returns></returns>
        public async Task<PlantType> AddPlant(FarmingComponentDTO farmingComponentDTO, long farmComponentId)
        {
            var plant = CopyFrom(farmingComponentDTO, farmComponentId);
            await plantRepository.AddNewPlant(plant);
            return plant;
        }

        /// <summary>
        /// Get All Plant exist in Database (NOT USE)
        /// </summary>
        /// <returns></returns>
        public List<PlantKB> GetAllPlant()
        {
            return plantRepository.GetAllPlant();
        }


        /// <summary>
        /// Get detail of a Plant
        /// </summary>
        /// <param name="farmComponentId"></param>
        /// <returns></returns>
        public List<PlantDetail> GetPlantDetail(long farmComponentId)
        {
            return plantRepository.GetPlantDetail(farmComponentId);
        }

        /// <summary>
        /// Copy data of a plant in Farm Component Command -> Plant Object
        /// </summary>
        /// <param name="farmingComponentDTO"></param>
        /// <param name="farmComponentId"></param>
        /// <returns></returns>
        private PlantType CopyFrom(FarmingComponentDTO farmingComponentDTO, long farmComponentId)
        {
            PlantType Plant = new PlantType()
            {
                PlantKBId = farmingComponentDTO.PlantKBId,
                Farm_ComponentId = farmComponentId,
                StartPlantDate = farmingComponentDTO.StartPlantDate,
                EndPlantDate = farmingComponentDTO.EndPlantDate,
                CurPeriod = 1,
                StartDayCurPeriod = 0
            };
            return Plant;
        }

        /// <summary>
        /// Ask System with current Farm Component, could user change current Period?
        /// </summary>
        /// <param name="farmComponentId"></param>
        /// <returns></returns>
        public ChangePeriodSignal AskChangePeriod(long farmComponentId)
        {
            try
            {
                var isLastPeriod = plantRepository.IsLastPeriod(farmComponentId);
                if (isLastPeriod)
                {
                    return new ChangePeriodSignal
                    {
                        Signal = SignalPeriod.IsLastPeroid
                    };
                }
                var isEnoughDay = plantRepository.IsEnoughDayToChangePeriod(farmComponentId);
                if (isEnoughDay)
                {
                    return new ChangePeriodSignal
                    {
                        Signal = SignalPeriod.IsAvailable
                    };
                }
                else
                {
                    return new ChangePeriodSignal
                    {
                        Signal = SignalPeriod.IsNotEnoughDay
                    };
                }
            }
            catch(ChangePeriodException e)
            {
                return new ChangePeriodSignal
                {
                    Signal = SignalPeriod.IsNotAvailable
                };
            }
        }

        /// <summary>
        /// Get detailed information of next Period in current Farm Component
        /// </summary>
        /// <param name="farmComponentId"></param>
        /// <returns></returns>
        public PeriodDetail GetNextPeriodDetail(long farmComponentId)
        {
            var userId = authenticationService.GetUserId();
            var isOwned = plantRepository.CheckFarmComponentWithUserId(userId, farmComponentId);
            if (isOwned)
            {
                return plantRepository.GetNextPeriod(farmComponentId);
            }
            else
            {
                throw new UnAuthorizedException("This Farm Component is not yours!");
            }
        }

        /// <summary>
        /// Change Period of current Farm Component to next Period
        /// </summary>
        /// <param name="farmComponentId"></param>
        /// <returns></returns>
        public async Task ChangeNextPeriod(long farmComponentId)
        {
            await plantRepository.ChangeNextPeriod(farmComponentId);
        }

        /// <summary>
        /// Check User is owing current Farm Component or not?
        /// </summary>
        /// <param name="farmComponentId"></param>
        /// <returns></returns>
        public bool CheckUserHaveFarmComponent(long farmComponentId)
        {
            var userId = authenticationService.GetUserId();
            return plantRepository.CheckFarmComponentWithUserId(userId,farmComponentId);
        }
    }
}

File: src/FarmingWeb/scripts/main/services/impl/plant.service.ts
import { Injectable } from "@angular/core";
import { Http, Headers, RequestOptions, Response } from "@angular/http";

import { PlantDetailModel } from './../../models/plant-detail.model';
import { PlantModel } from './../../models/plant.model';
import { ChangePeriodSignal } from './../../models/change-period-signal.model';
import { PeriodDetail } from './../../models/period-detail.model';

import { IPlantService } from './../interface/plant-service.interface';

import { AppSetting } from './../../../app/app.setting';

const headers: Headers = new Headers({ "Content-Type": "application/json" });
const options: RequestOptions = new RequestOptions({ headers: headers });

//this service is used for some logics related to Plant
@Injectable()
export class PlantService implements IPlantService {
    constructor(
        protected http: Http
    ) { }

    //get Plant Detail of a FarmComponent
    getPlantDetail(farmComponentId: number): Promise<PlantDetailModel[]> {
        return new Promise<any>((resolve: any, reject: any) => {
            this.http.get(AppSetting.API_ENDPOINT + '/plants/detail/' + farmComponentId).subscribe(
                (data: Response) => {
                    let listPlantDetail = <PlantDetailModel[]>data.json();
                    resolve(listPlantDetail);
                },
                (error: any) => {
                    reject(error);
                }
            );
        });
    }

    //get All Plants in Database
    getAllPlant(): Promise<PlantModel[]> {
        return new Promise<any>((resolve: any, reject: any) => {
            this.http.get(AppSetting.API_ENDPOINT + '/plants/').subscribe(
                (data: Response) => {
                    let listPlantDetail = <PlantModel[]>data.json();
                    resolve(listPlantDetail);
                },
                (error: any) => {
                    reject(error);
                }
            );
        });
    }

    //ask Server plant in FarmComponent can change period now?
    askChangePeriod(farmComponentId: number): Promise<ChangePeriodSignal> {
        return new Promise<ChangePeriodSignal>((resolve: any, reject: any) => {
            this.http.get(AppSetting.API_ENDPOINT + '/plants/changePeriod/' + farmComponentId).subscribe(
                (data: Response) => {
                    let signal = <ChangePeriodSignal>data.json();
                    resolve(signal);
                },
                (error: any) => {
                    reject(error);
                }
            );
        });
    }

    //get next period Detail
    getNextPeriod(farmComponentId: number): Promise<PeriodDetail> {
        return new Promise<PeriodDetail>((resolve: any, reject: any) => {
            this.http.get(AppSetting.API_ENDPOINT + '/plants/nextPeriod/' + farmComponentId).subscribe(
                (data: Response) => {
                    let periodDetail = <PeriodDetail>data.json();
                    resolve(periodDetail);
                },
                (error: any) => {
                    reject(error);
                }
            );
        });
    }

    //Change period to next period of a Plant
    changePeriod(farmComponentId: number): Promise<void> {
        return new Promise<void>((resolve: any, reject: any) => {
            this.http.post(AppSetting.API_ENDPOINT + '/plants/changePeriod/' + farmComponentId, null, options).subscribe(
                (data: Response) => {
                    resolve();
                },
                (error: any) => {
                    reject(error);
                }
            );
        });
    }
}